onnxruntime/core/providers/cuda/tensor/tile_impl.cu

```diff
@@ -7,140 +7,178 @@
-    output_index = r;
-    int in_coord = fdm_input_shape[dim].mod(out_coord);
-    input_index += input_strides[dim] * in_coord;
   }
-  output_data[id] = input_data[input_index];
-}
 
-template <typename T>
-__global__ void _TileKernel(
-    const size_t shape_rank,
-    const TArray<fast_divmod> fdm_input_shape,
-    const TArray<int64_t> input_strides,
-    const T* input_data,
-    const TArray<fast_divmod> fdm_output_strides,
-    T* output_data,
-    const CUDA_LONG N) {
-  CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
-  CUDA_LONG input_index = 0;
-  CUDA_LONG output_index = id;
-  for (int dim = 0; dim < shape_rank; ++dim) {
-    int out_coord, r;
-    fdm_output_strides[dim].divmod(output_index, out_coord, r);
-    output_index = r;
-    int in_coord = fdm_input_shape[dim].mod(out_coord);
-    input_index += input_strides[dim] * in_coord;
+  id = start;
+#pragma unroll
+  for (int i = 0; i < num_elements_per_thread; ++i) {
+    if (id < N) {
+      output_data[id] = value[i];
+      id += num_threads_per_block;
+    }
   }
-  output_data[id] = input_data[input_index];
 }
 
 template <typename T>
-void TileImpl(
-    cudaStream_t stream,
-    const size_t shape_rank,
-    const TArray<fast_divmod>& fdm_input_shape,
-    const TArray<int64_t>& input_stride,
-    const T* input_data,
-    const TArray<fast_divmod>& fdm_output_strides,
-    T* output_data,
-    const size_t N) {
-  int blocksPerGrid = (int)(ceil(static_cast<float>(N) / GridDim::maxThreadsPerBlock));
-  if (shape_rank > MAX_DIMS) {
-    _TileKernel<T><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
-        shape_rank, fdm_input_shape, input_stride, input_data,
-        fdm_output_strides, output_data, (CUDA_LONG)N);
-  } else {
-    _UnRolledTileKernel<T><<<blocksPerGrid, GridDim::maxThreadsPerBlock, 0, stream>>>(
-        shape_rank, fdm_input_shape, input_stride, input_data,
-        fdm_output_strides, output_data, (CUDA_LONG)N);
-  }
+void TileImpl(cudaStream_t stream, const size_t shape_rank, const TArray<fast_divmod>& fdm_input_shape,
+              const TArray<int64_t>& input_stride, const T* input_data, const TArray<fast_divmod>& fdm_output_strides,
+              T* output_data, const size_t N) {
+  int blocksPerGrid = static_cast<int>(CeilDiv(N, num_threads_per_block * num_elements_per_thread));
+  _UnRolledTileKernel<T><<<blocksPerGrid, num_threads_per_block, 0, stream>>>(shape_rank, fdm_input_shape, input_stride,
+                                                                              input_data, fdm_output_strides,
+                                                                              output_data, static_cast<CUDA_LONG>(N));
 }
 
 template <typename T>
-__global__ void _TileMemcpyKernel(
-    const T* input_data,
-    const size_t num_input_elements,
-    T* output_data,
-    const size_t N) {
+__global__ void _TileMemcpyKernel(const T* input_data, T* output_data, const CUDA_LONG N, const size_t repeats) {
```

Comment:
Just curious: What happens if we just use `__restrict__` for `input_data` ? That should enable some caching mitigating the ill-effects of the multiple global memory reads ? It is probably not as good as this approach, I am guessing.

Reply:
Had a quick try, added this both to the old code and new code, but didn't see any perf gain. I am thinking NVCC already tries to use the const memory caching as possible as it can, no matter wee have the __restrict__ in the code.